### PR TITLE
[dev] Use expose settings when generating ingress rules

### PR DESCRIPTION
**Pull request by achilleasa**
## Description of change

This PR refactors the ingress rule generation logic in the firewall worker so that it takes into account any per-endpoint expose settings when generating ingress rules for the various substrates. The expose settings can be manipulated by the `juju expose` and `juju unexpose` commands (see https://github.com/juju/juju/pull/11984 for more details) as well as via the overlay section in bundles (see https://github.com/juju/juju/pull/11992).

Prior to this change, the firewall worker would check whether an application is exposed, and if so, it would generate an ingress rule for each open port rage using `0.0.0.0/0` as the source CIDR. This PR changes the ingress generation logic for exposed applications as follows:

For each **explicit endpoint** expose setting:
  - Calculate the effective set of source CIDRs by taking the union of:
     - any operator-provided CIDRS (e.g. `juju expose apache2 --endpoint foo --to-cidrs 10.0.0.0/24`) **for the endpoint** 
     - the CIDRs that correspond to the subnets that are associated with each space that the **endpoint is exposed to** (e.g. `juju expose apache2 --endpoint foo --to-spaces external`).
  - Calculate the set of effective port ranges that the charm has opened to this endpoint by taking the union of any ports explicitly opened for that endpoint (e.g. `open-port 1337/tcp --endpoints foo`) and any ports opened for **all** endpoints (e.g. `open-port 80/tcp` -- what you would expect from a 2.8 controller).
 - Generate an ingress rule for each port range and source CIDR combination.

For the expose **all** endpoints setting (e.g. `juju expose apache2 --to-cidrs 10.0.0.0/24,192.168.0.0/24 --to-spaces alpha`):
  - Calculate the effective set of source CIDRs by taking the union of:
     - any operator-provided CIDRS (e.g. `juju expose apache2 --to-cidrs 10.0.0.0/24`) for the **wildcard** endpoint 
     - the CIDRs that correspond to the subnets that are associated with each space that the **wildcard** endpoint is exposed to (e.g. `juju expose apache2 --to-spaces external`).
  - Calculate the set of effective port ranges that the charm has opened to this endpoint by adding the ports explicitly opened for **each** endpoint **if and only if** an application expose setting does not exist for that particular endpoint.
 - Generate an ingress rule for each port range and source CIDR combination.

To clarify the port selection logic for the expose-all-endpoints case, here is an example:

- Apache2 opens:
   - Port 80 for all endpoints
   - Port 1337 for the website endpoint
   - Port 8080 for the logs endpoint
- Expose apache2 --to-cidr 0.0.0.0/0
- Expose apache2 --endpoint website --to-cidr 10.0.0./24

When calculating the ports for the expose-all-endpoints, the set will include:
- port 80 (opened for all endpoints)
- port 8080 (matches the expose all endpoints and **no expose settings exist** for the `logs` endpoint)

Note that port 1337 is not included because there is an explicit override for it (the relevant ingress rules will be generated via the **explicit** endpoint rules).

## API changes introduced by this PR

To implement this functionality, new API methods had to be introduced to the firewall facade. As the API has _already been bumped in develop, no further bump was required_. The following new calls have been added (see individual commits below):
- `ExposeInfo`: returns the expose flag **and** the per-endpoint expose settings (supersedes `GetExposed`)
- `SpaceInfo`: a.k.a "how can I use network.SpaceInfos outside the controller"? Returns a SpaceInfos list, optionally filtered by a set of space IDs.
- `OpenedMachinePortRanges`: (client only; actual method introduced in https://github.com/juju/juju/pull/11956)
- `WatchSubnets`:  used by the worker to detect space topology changes (e.g. subnets moving around) and regenerate ingress rules for any applications that include at least one expose-to-space setting. Note: the original watcher (doesn't seem to be referenced anywhere except a commented section in WatchEgressAddresses) used a lifecycle watcher but I changed it to a collection watcher.

As the worker will be restarted together with the controller after an upgrade, it is possible to get rid of the old facade methods (client/controller) that got superseded by the above methods without the risk to break anything. Just in case we ever need to revert this change in the future, the removal code (and facade schema changes) have been isolated in a standalone commit (https://github.com/juju/juju/pull/12001/commits/9bfb456507d38ea65b3abeb43ad9f339a0ee774d)

## QA steps

Try this on a provider supporting firewalls (e.g. aws, azure, gce or openstack). I have already tried the following steps on aws and azure.

```console
$ juju bootstrap azure test --no-gui
$ juju deploy apache2
$  juju run --unit apache2/0 "open-port 80/tcp"
$ juju run --unit apache2/0 "open-port 1337/tcp --endpoints website"
$ juju run --unit apache2/0 "opened-ports --endpoints"
80/tcp (*)
1337/tcp (website)
```

Run the following commands and inspect the firewall rules in the substrate's UI...

### Expose all ports

``` console
$ juju expose apache2
```

image

### Override expose rules for specific endpoint

```console
$ juju expose apache2 --endpoints website --to-cidrs 10.0.0.0/24
```

image

Note: as port 80 opened for all endpoints (including `website), it will get an ingress rule with sources 0.0.0.0/0 **and** 10.0.0.0/24 (from the above juju expose).

### Unexpose endpoint

```console
$ juju unexpose apache2 --endpoints website
```

image

### Expose to space

```console
$ juju unexpose apache2
$ juju expose apache2 --endpoints website --to-spaces alpha
```

image

Note: port 80 is opened for all endpoints and port 1337 is only opened for the website endpoint; so you should get one rule for each using the CIDRs for the alpha space subnets.

### Other things to try

Do an expose to space operation and then move a subnet to a different space; verify that the ingress rules have changed to include the updated CIDRs.

## Pending work

- juju bundle diff (show exposed endpoints)
- juju show application (show exposed endpoints)
- pylibjuju

## Documentation changes

I have created a card to revise the documentation for this stream of work.

### CODE REVIEW

**Comment by achilleasa** [comment]
I need to do a similar fix as with openstack for the joyent provider; will take a look tomorrow.

File: apiserver/facades/controller/firewaller/firewaller.go
Comment (achilleasa) [inline apiserver/facades/controller/firewaller/firewaller.go:650]: @manadart This is the one we were talking about. All the marshal/unmarshal logic is encapsulated in the params package so we can easily clone this into any other facades that need this functionality.

File: apiserver/facades/controller/firewaller/firewaller.go
Comment (wallyworld) [inline apiserver/facades/controller/firewaller/firewaller.go:705]: Not for this PR but I wish we had a proper server side business logic layer where we'd have a WatchSubnets(ids ...[]string) and pass in ids extracted from parsed tags and do all the filter setup etc in that business logic layer.

File: apiserver/params/applications.go
Comment (wallyworld) [inline apiserver/params/applications.go:569]: Would it benefit to use "*" as the match everything string? For clarity when doing db dumps etc?
Maybe not worth it. But if not already done, I assume there's a const defined?

File: apiserver/params/network.go
Comment (wallyworld) [inline apiserver/params/network.go:1262]: We've previously defined these in the relevant apiserver facade package but I think here is better actually.

File: worker/firewaller/firewaller.go
Comment (wallyworld) [inline worker/firewaller/firewaller.go:356]: expose endpoints

**Review by wallyworld — APPROVED** [review]
I haven't done a deep dive into the detailed logic to combine port ranges etc but it all seems nicely put together.
+1 pending a second review from Joe.
We'll need to consider impact on the new k8s firewaller also where work is currently being done to open ports on the service.

File: apiserver/params/applications.go
Comment (achilleasa) [reply apiserver/params/applications.go:569]: The `""` was chosen so we are consistent with endpoint bindings (the `""` represents the default space for all endpoints) both in persisted state and in bundles. Also, the same pattern is used in the openedPorts collection to indicate that a port is opened for all endpoints (present and future -- see https://github.com/juju/juju/pull/11863 for more details about the wildcard endpoint wrt. port ranges). 

Most of the tests use the `allEndpoints` const but the logic typically checks for `""` (though I have sprinkled comments around those locations to explain what this represents)

File: apiserver/params/network.go
Comment (achilleasa) [reply apiserver/params/network.go:1262]: As mentioned in the extended commit log for this change, I moved them there because we _may_ need to implement `SpaceInfos` in other facades in the future (so we can consistently use SpaceInfos across the codebase) and having the marshal/unmarshal logic in `params` makes it very easy to clone the API method (also avoids the `networkingcommon` package pattern which I believe we are trying to get rid of eventually)

**Comment by achilleasa** [comment]
> We'll need to consider impact on the new k8s firewaller also where work is currently being done to open ports on the service.

@wallyworld I was planning to ask about the k8s-related work for this feature in this week's cross-team meeting as it is still an open question in the spec.

File: api/firewaller/firewaller_test.go
Comment (achilleasa) [reply api/firewaller/firewaller_test.go:293]: I left the check in the client just to be safe but it shouldn't be possible to have a version mismatch between the controller and this facade in particular as you automatically get the new client/facade versions when you upgrade the controller. So I didn't feel the need to add a test for older facade versions.

**Comment by achilleasa** [comment]
!!build!! (windows box aborted build)

**Comment by jameinel** [comment]
The documentation from the PR should probably be turned into a discourse post about how to utilize this functionality.

File: apiserver/params/applications.go
Comment (jameinel) [reply apiserver/params/applications.go:569]: I wish we had used "*" for the "default binding" in the beginning. It would make things much easier to talk about.
I wonder if we could sneak it into a 3.0 ...

File: apiserver/params/network.go
Comment (jameinel) [inline apiserver/params/network.go:1296]: I wonder if there is a cleaner structure for this. A V3 holding a V2 holding a V1 doesn't inspire me, and is sort of the inverse of how we do Facade versioning. (As we go V4, etc, it just gets worse.)
Doesn't need to change here, but something we should consider.

File: core/network/portrange_test.go
Comment (jameinel) [inline core/network/portrange_test.go:486]: but these aren't the same length (according to descr)

I don't know if the code is incorrect, but the test data doesn't match the description.

File: worker/firewaller/firewaller.go
Comment (jameinel) [inline worker/firewaller/firewaller.go:7]: I got to here, but I want to give this part more focused attention, I will try to look at this a bit later. (dinner time)

File: core/network/portrange_test.go
Comment (achilleasa) [reply core/network/portrange_test.go:486]: The description refers to the length of the GroupedPortRanges _keys_ and not their values. It is meant to test the slow path (i.e. they have same number of keys so we need to iterate and compare both keys and their contents). I can update the description if it seems confusing.

File: worker/firewaller/firewaller_test.go
Comment (manadart) [inline worker/firewaller/firewaller_test.go:1421]: Typo.

**Review by manadart — COMMENTED** [review]
The code looks good. I think there are parts that could have been proposed separately to keep it leaner, but that's OK.

I have not been able to QA on MicroStack due to having tribulations with my setup.

There is one thing I think we need to decide upon. Take this test for example (quick and dirty mod of `TestExposedApplicationWithExposedEndpointsWhenSpaceTopologyChanges`):
```go
func (s *InstanceModeSuite) TestExposedApplicationWithExposedEndpointsWhenSpaceDeleted(c *gc.C) {
	// Create two spaces and add a subnet to each one
	sp1, err := s.State.AddSpace("space1", "sp-1", nil, false)
	c.Assert(err, jc.ErrorIsNil)
	sub1, err := s.State.AddSubnet(network.SubnetInfo{
		ID:        "subnet-1",
		CIDR:      "192.168.0.0/24",
		SpaceID:   sp1.Id(),
		SpaceName: sp1.Name(),
		IsPublic:  false,
	})
	c.Assert(err, jc.ErrorIsNil)

	sp2, err := s.State.AddSpace("space2", "sp-2", nil, true)
	c.Assert(err, jc.ErrorIsNil)
	_, err = s.State.AddSubnet(network.SubnetInfo{
		ID:        "subnet-2",
		CIDR:      "192.168.1.0/24",
		SpaceID:   sp2.Id(),
		SpaceName: sp2.Name(),
		IsPublic:  false,
	})
	c.Assert(err, jc.ErrorIsNil)

	fw := s.newFirewaller(c)
	defer statetesting.AssertKillAndWait(c, fw)

	app := s.AddTestingApplication(c, "wordpress", s.charm)

	u, m := s.addUnit(c, app)
	inst := s.startInstance(c, m)
	// Open port 80 for all endpoints
	mustOpenPortRanges(c, s.State, u, allEndpoints, []network.PortRange{
		network.MustParsePortRange("80/tcp"),
	})

	// Expose app to space-1
	err = app.MergeExposeSettings(map[string]state.ExposedEndpoint{
		allEndpoints: {
			ExposeToSpaceIDs: []string{sp1.Id()},
		},
	})
	c.Assert(err, jc.ErrorIsNil)

	s.assertIngressRules(c, inst, m.Id(), firewall.IngressRules{
		// We expect to see port 80 use the subnet-1 CIDR
		firewall.NewIngressRule(network.MustParsePortRange("80/tcp"), "192.168.0.0/24"),
	})

	// Simulate the deletion of a space, with subnets moving back to alpha.
	moveOps := sub1.UpdateSpaceOps(network.AlphaSpaceId)
	deleteOps := sp1.RemoveSpaceOps()
	c.Assert(s.State.ApplyOperation(modelOp{append(moveOps, deleteOps...)}), jc.ErrorIsNil)

func (s *InstanceModeSuite) TestExposedApplicationWithExposedEndpointsWhenSpaceDeleted(c *gc.C) {
	// Create two spaces and add a subnet to each one
	sp1, err := s.State.AddSpace("space1", "sp-1", nil, false)
	c.Assert(err, jc.ErrorIsNil)
	sub1, err := s.State.AddSubnet(network.SubnetInfo{
		ID:        "subnet-1",
		CIDR:      "192.168.0.0/24",
		SpaceID:   sp1.Id(),
		SpaceName: sp1.Name(),
		IsPublic:  false,
	})
	c.Assert(err, jc.ErrorIsNil)

	sp2, err := s.State.AddSpace("space2", "sp-2", nil, true)
	c.Assert(err, jc.ErrorIsNil)
	_, err = s.State.AddSubnet(network.SubnetInfo{
		ID:        "subnet-2",
		CIDR:      "192.168.1.0/24",
		SpaceID:   sp2.Id(),
		SpaceName: sp2.Name(),
		IsPublic:  false,
	})
	c.Assert(err, jc.ErrorIsNil)

	fw := s.newFirewaller(c)
	defer statetesting.AssertKillAndWait(c, fw)

	app := s.AddTestingApplication(c, "wordpress", s.charm)

	u, m := s.addUnit(c, app)
	inst := s.startInstance(c, m)
	// Open port 80 for all endpoints
	mustOpenPortRanges(c, s.State, u, allEndpoints, []network.PortRange{
		network.MustParsePortRange("80/tcp"),
	})

	// Expose app to space-1
	err = app.MergeExposeSettings(map[string]state.ExposedEndpoint{
		allEndpoints: {
			ExposeToSpaceIDs: []string{sp1.Id()},
		},
	})
	c.Assert(err, jc.ErrorIsNil)

	s.assertIngressRules(c, inst, m.Id(), firewall.IngressRules{
		// We expect to see port 80 use the subnet-1 CIDR
		firewall.NewIngressRule(network.MustParsePortRange("80/tcp"), "192.168.0.0/24"),
	})

	// Simulate the deletion of a space, with subnets moving back to alpha.
	moveOps := sub1.UpdateSpaceOps(network.AlphaSpaceId)
	deleteOps := sp1.RemoveSpaceOps()
	c.Assert(s.State.ApplyOperation(modelOp{append(moveOps, deleteOps...)}), jc.ErrorIsNil)

	// The space we were exposed to was deleted. Keep the old rules?
	s.assertIngressRules(c, inst, m.Id(), firewall.IngressRules{
		firewall.NewIngressRule(network.MustParsePortRange("80/tcp"), "192.168.0.0/24"),
	})
}
```

It fails with:
```
firewaller_test.go:1481:
    // Check that worker picked up the change and updated the rules
    s.assertIngressRules(c, inst, m.Id(), firewall.IngressRules{
        // We expect to see port 80 use subnter-{1,2} CIDRs
        firewall.NewIngressRule(network.MustParsePortRange("80/tcp"), "192.168.0.0/24", "192.168.1.0/24"),
    })
firewaller_test.go:130:
    c.Fatalf("timed out: expected %q; got %q", expected, got)
... Error: timed out: expected ["80/tcp from 192.168.0.0/24,192.168.1.0/24"]; got ["80/tcp"]
```

This would indicate that deleting a space with endpoints exposed to it actually makes those rules expose to all subnets, which wouldn't be what we want.

The same would go for a space that have all subnets moved out of it.

File: provider/joyent/local_test.go
Comment (achilleasa) [reply provider/joyent/local_test.go:94]: The joyent bits will get purged in https://github.com/juju/juju/pull/12006

**Review by manadart — APPROVED** [review]
Based on discussions regarding a follow-up, this is OK to land.

As I understand it, we will no longer imply exposure to all subnets based on an empty CIDR list. Instead we will explicitly represent 0.0.0.0/::1, and an empty list is really no exposure. This will require an upgrade/migration step.

File: worker/firewaller/firewaller_test.go
Comment (achilleasa) [reply worker/firewaller/firewaller_test.go:1421]: Will fix in the followup PR.

**Comment by achilleasa** [comment]
$$merge$$

**Pull request merged**
